weather_dl/download_pipeline/fetcher.py

```diff
@@ -44,29 +44,122 @@ class Fetcher(beam.DoFn):
+        )
+
+
+class EachPartition(beam.DoFn):
+    def process(self, element) -> t.Iterator[t.Tuple[t.Any, str]]:
+        # element: Tuple[Tuple[str, int], Iterator[Config]]
+        """Execute download requests one-by-one."""
+        (subsection, request_idx), partitions = element
+        worker_name = f'{subsection}.{request_idx}'
+        logger.info(f"[{worker_name}] Starting requests...")
+        for partition in partitions:
+            beam.metrics.Metrics.counter('Fetcher', subsection).inc()
+            yield (partition, worker_name)
+
+
+@dataclasses.dataclass
+class FetchData(beam.DoFn):
+    client_name: str
+    manifest: Manifest = NoOpManifest(Location('noop://in-memory'))
+    store: t.Optional[Store] = None
+
     @retry_with_exponential_backoff
-    def upload(self, src: io.FileIO, dest: str) -> None:
-        """Upload blob to cloud storage, with retries."""
-        with self.store.open(dest, 'wb') as dest_:
-            shutil.copyfileobj(src, dest_, WRITE_CHUNK_SIZE)
+    def fetch(self, client: Client, dataset: str, selection: t.Dict) -> t.Dict:
+        """Fetch data from download client, with retries."""
+        return client.fetch(dataset, selection)
+
+    def process(self, element) -> t.Iterator[t.Tuple[t.Dict, Config, str, str, str]]:
+        """Fetch data from client."""
+        config, worker_name = element
+
+        if not config:
+            return
+
+        if skip_partition(config, self.store):
+            return
+
+        client = CLIENTS[self.client_name](config)
+        target = prepare_target_name(config)
+
+        with self.manifest.transact(config.selection, target, config.user_id):
+            with tempfile.NamedTemporaryFile() as temp:
+                logger.info(f'[{worker_name}] Fetching data for {target!r}.')
+                result = self.fetch(client, config.dataset, config.selection)
+                yield (result, config, worker_name, temp.name, target)
+
+
+@dataclasses.dataclass
+class DownloadData(beam.DoFn):
+    client_name: str
+    manifest: Manifest = NoOpManifest(Location('noop://in-memory'))
+    store: t.Optional[Store] = None
+
+    @retry_with_exponential_backoff
+    def download(self, client: Client, dataset: str, result: t.Dict, dest: str) -> None:
+        """Download data from download client, with retries."""
+        client.download(dataset, result, dest)
+
+    def process(self, element) -> t.Iterator[t.Tuple[Config, str, str, str]]:
+        """Download data from a client to a temp file."""
+        result, config, worker_name, temp_name, target = element
+
+        client = CLIENTS[self.client_name](config)
+
+        with self.manifest.transact(config.selection, target, config.user_id):
+            logger.info(f'[{worker_name}] Downloading data for {target!r}.')
+            self.download(client, config.dataset, result, temp_name)
+            yield (config, worker_name, temp_name, target)
+
+
+@dataclasses.dataclass
+class Retrieve(beam.DoFn):
```

Comment:
Nit: maybe it should be `RetrieveData` to follow the naming conventions of the other classes?